libsolidity/interface/CompilerStack.cpp

```diff
@@ -507,21 +507,44 @@ string CompilerStack::applyRemapping(string const& _path, string const& _context
 			return false;
 		return std::equal(_a.begin(), _a.end(), _b.begin());
 	};
+	// Try to find whether _a is a closer match for context _reference than _b
+	// Defaults to longest prefix in case of a tie.
+	auto isClosestContext = [](string const& _a, string const& _b, string const& _reference)
+	{
+		int a = _reference.compare(_a);
+		int b = _reference.compare(_b);
+		if (a == 0)
+			return true;
+		else if (b == 0)
+			return false;
+		else if (abs(a) == abs(b)) {
+			return a > 0;
+		}
+		return abs(a) < abs(b);
+	};
 
+	using filepath = boost::filesystem::path;
+	filepath context(_context);
 	size_t longestPrefix = 0;
 	string longestPrefixTarget;
+	string currentClosestContext;
+	string referenceContext = context.parent_path().generic_string();
 	for (auto const& redir: m_remappings)
 	{
+		filepath redirContext(redir.context);
 		// Skip if we already have a closer match.
-		if (longestPrefix > 0 && redir.prefix.length() <= longestPrefix)
+		if (longestPrefix > 0 && redir.prefix.length() < longestPrefix)
```

Comment:
I have the impression that context-length should be more important than prefix length. As an example, suppose we have the three remappings

```
dir:std=/usr/share/std
dir:std/list=/usr/share/my_special_list
dir/sub:std=/usr/share/my_special_std/
```

This means that the default is to use some std library, but replace its list implementation by a different one. Inside the subdirectory, we use a completely different std library. I would say that it does not make much sense anymore to keep the "deeper" redirect, because even the meaning of "std" might be different inside `dir/sub`, i.e. it might be a completely different interface where `my_special_list` would not fit anymore.